File: dashboard/src/lib/types/analytics.ts
// Analytics types matching gateway models

// Re-export TokenRow from token.ts for consistency
export type { TokenRow } from "./token"

export interface AnalyticsSummary {
  total_requests: number
  success_count: number
  error_count: number
  avg_latency: number
  total_cost: number
  total_tokens: number
}

export interface AnalyticsTimeseriesPoint {
  bucket: string
  request_count: number
  error_count: number
  cost: number
  lat: number
}

export interface LatencyStat {
  p50: number
  p90: number
  p99: number
  avg: number
}

export interface TokenSummary {
  token_id: string
  total_requests: number
  errors: number
  avg_latency_ms: number
  last_active: string | null
}

export interface SpendCap {
  token_id: string
  token_name: string
  spend_used_usd: number | null
  spend_cap_usd: number
}

export interface UpstreamStatus {
  name: string
  healthy: boolean
  last_check: string
  latency_ms: number | null
  error: string | null
}

// Provider analytics types
export interface ModelUsageStat {
  model: string
  request_count: number
  cost_usd: number
}

export interface ProviderSpendStat {
  provider: string
  spend_usd: number
  rate_per_1k: number
}

export interface ProviderLatencyStat {
  provider: string
  latency_ms: number
}

// Traffic analytics types (Traffic Tab)
export interface TrafficTimeseriesPoint {
  bucket: string
  total_count: number
  passed_count: number
  throttled_count: number
  blocked_count: number
  hitl_paused_count: number
}

export interface LatencyTimeseriesPoint {
  bucket: string
  p50: number
  p90: number
  p99: number
}

export interface AuditLogRow {
  request_id: string
  token_id: string
  model: string | null
  method: string
  response_latency_ms: number
  upstream_status: number | null
  policy_result: string
  estimated_cost_usd: number | null
  created_at: string
}

// Cost analytics types (Cost Tab)

export interface BudgetHealthStatus {
  tokens_above_80_percent: number
  tokens_without_cap: number
  total_tokens: number
}

export interface SpendTimeseriesPoint {
  bucket: string
  dimension: string
  spend_usd: number
  request_count: number
}

export interface CostEfficiencyPoint {
  bucket: string
  model: string
  cost_per_1k_tokens: number
}

export interface BudgetBurnRate {
  days_elapsed: number
  days_remaining: number
  budget_usd: number
  spent_usd: number
  percent_used: number
  needed_per_day: number
  actual_per_day: number
  on_track: boolean
}

export interface TokenSpendWithCap {
  token_id: string
  token_name: string
  provider: string
  total_spend_usd: number
  spend_cap_usd: number | null
  percent_cap_used: number | null
  request_count: number
  cost_per_1k: number
}

// Spend breakdown types
export interface SpendByDimension {
  dimension: string
  total_cost_usd: number
  request_count: number
  total_prompt_tokens: number
  total_completion_tokens: number
}

// ── Users & Tokens Analytics Types ─────────────────────────────────────

/// User growth timeseries point for the Users & Tokens tab.
export interface UserGrowthPoint {
  bucket: string
  new_users: number
  cumulative_users: number
}

/// Engagement tier breakdown for the Users & Tokens tab.
export interface EngagementTiersResponse {
  power_users: number
  regular_users: number
  light_users: number
  total_users: number
}

/// A token that is at or approaching rate limit.
export interface RateLimitedToken {
  token_name: string
  percent: number
}

/// Token alerts for the Users & Tokens tab.
export interface TokenAlertsResponse {
  active_tokens: number
  token_limit: number | null
  tokens_at_rate_limit: number
  rate_limited_tokens: RateLimitedToken[]
}

/// Requests per user timeseries point for the Users & Tokens tab.
export interface RequestsPerUserPoint {
  bucket: string
  user_count: number
  request_count: number
  avg_per_user: number
}

// ── Cache Analytics Types ─────────────────────────────────────

/// Cache summary statistics for the Cache tab ribbon.
export interface CacheSummaryStats {
  hit_rate: number
  cost_avoided_usd: number
  cache_size_bytes: number
  top_cached_model: string | null
}

/// Cache hit rate timeseries point for the Cache tab chart.
export interface CacheHitRatePoint {
  bucket: string
  hit_rate: number
  hit_count: number
  total_count: number
}

/// Top cached query entry for the Cache tab table.
export interface CachedQueryRow {
  query_hash: string
  model: string | null
  hits: number
  last_hit_at: string
  cache_age_seconds: number
  expires_in_seconds: number | null
}

/// Model-level cache efficiency for the Cache tab.
export interface ModelCacheEfficiency {
  model: string
  hit_rate: number
  total_requests: number
  cache_hits: number
}

/// Cache latency comparison for the Cache tab.
export interface CacheLatencyComparison {
  cached_latency_ms: number
  uncached_latency_ms: number
  speedup_factor: number
  cached_sample_count: number
  uncached_sample_count: number
}

// ── Model Analytics Types (Models Tab) ─────────────────────────────────────

/// Model usage timeseries point for the Models tab.
export interface ModelUsageTimeseriesPoint {
  bucket: string
  model: string
  value: number
}

/// Error rate by model for the Models tab.
export interface ModelErrorRate {
  model: string
  error_rate: number
  total_requests: number
  error_count: number
}

/// Latency by model for the Models tab.
export interface ModelLatencyStat {
  model: string
  avg_latency_ms: number
  p50: number
  p90: number
  p99: number
  sample_count: number
}

/// Combined model stats for the Models tab table.
export interface ModelStatsRow {
  model: string
  total_tokens: number
  total_cost_usd: number
  error_rate: number
  avg_latency_ms: number
  request_count: number
}

/// Cost vs latency scatter point for the Models tab bubble chart.
export interface CostLatencyScatterPoint {
  model: string
  avg_cost_per_request: number
  avg_latency_ms: number
  total_spend_usd: number
  total_requests: number
}

/// A/B test lift metrics for the Models tab.
export interface ABTestLiftResponse {
  experiment_id: string
  experiment_name: string
  champion: ABTestVariant
  challenger: ABTestVariant
  cost_delta_percent: number
  quality_metrics: QualityMetricDelta[]
}

/// A/B test variant info.
export interface ABTestVariant {
  model: string
  weight_percent: number
}

/// Quality metric delta for A/B test lift.
export interface QualityMetricDelta {
  metric: string
  delta: string
}

/// Experiment summary for A/B testing.
export interface ExperimentSummary {
  experiment_name: string
  variant_name: string
  total_requests: number
  avg_latency_ms: number
  total_cost_usd: number
  avg_tokens: number
  error_count: number
}

// ── Security Analytics Types (Security Tab) ─────────────────────────────────────

/// Security KPI summary for the Security tab ribbon.
export interface SecuritySummaryStats {
  /// Total PII redactions
  pii_redactions: number
  /// Total guardrail blocks
  guardrail_blocks: number
  /// Total shadow mode violations
  shadow_violations: number
  /// External guardrail blocks
  external_blocks: number
}

/// Guardrail trigger count by category for the Security tab.
export interface GuardrailTriggerStat {
  /// Guardrail category
  category: string
  /// Number of triggers
  count: number
}

/// PII breakdown by pattern type for the Security tab.
export interface PiiBreakdownStat {
  /// PII pattern type
  pattern: string
  /// Number of redactions
  count: number
}

/// Policy action count for the Security tab.
export interface PolicyActionStat {
  /// Policy action type
  action: string
  /// Number of times applied
  count: number
}

/// Shadow mode policy with violation stats for the Security tab.
export interface ShadowPolicyStat {
  /// Policy name
  policy_name: string
  /// Number of shadow violations
  violations: number
  /// Token with most violations
  top_token: string
  /// Policy status
  status: string
}

/// Data residency stats for the Security tab.
export interface DataResidencyStats {
  /// Percentage of EU-routed requests
  eu_percent: number
  /// Percentage of US-routed requests
  us_percent: number
}

// ── HITL Analytics Types (HITL Tab) ─────────────────────────────────────

/// HITL KPI summary for the HITL tab ribbon.
export interface HitlSummaryStats {
  /// Number of pending approval requests
  pending_count: number
  /// Average wait time in seconds for completed approvals
  avg_wait_seconds: number
  /// Approval rate as percentage (0-100)
  approval_rate: number
}

/// HITL volume timeseries point for the HITL tab chart.
export interface HitlVolumePoint {
  /// Time bucket
  bucket: string
  /// Count of approved requests
  approved_count: number
  /// Count of rejected requests
  rejected_count: number
  /// Count of expired requests (proxy for escalated)
  expired_count: number
  /// Count of pending requests
  pending_count: number
}

/// HITL latency stats for the SLA card.
export interface HitlLatencyStats {
  /// P50 approval time in seconds
  p50_seconds: number
  /// P90 approval time in seconds
  p90_seconds: number
  /// P99 approval time in seconds
  p99_seconds: number
  /// Average approval time in seconds
  avg_seconds: number
}

/// Rejection reason breakdown for the HITL tab.
export interface RejectionReason {
  /// Reason for rejection
  reason: string
  /// Percentage of total rejections
  percentage: number
}

/// Approval request from the gateway.
export interface ApprovalRequest {
  id: string
  token_id: string
  project_id: string
  idempotency_key: string | null
  request_summary: {
    method: string
    path: string
    agent?: string
    upstream?: string
    body_preview?: string
  } | null
  status: "pending" | "approved" | "rejected" | "expired"
  reviewed_by: string | null
  reviewed_at: string | null
  expires_at: string
  created_at: string
}

// ── Error Analytics Types (Errors Tab) ─────────────────────────────────────

/// Error KPI summary for the Errors tab ribbon.
export interface ErrorSummaryStats {
  /// Total error count
  total_errors: number
  /// Error rate as percentage (0-100)
  error_rate: number
  /// Circuit breaker trip count
  circuit_breaker_trips: number
  /// Rate limit hit count
  rate_limit_hits: number
  /// Most common error type
  top_error_type: string | null
  /// Prior period total errors for delta
  prior_total_errors: number
  /// Prior period total requests for delta
  prior_total_requests: number
}

/// Error timeseries point for the Errors tab chart.
export interface ErrorTimeseriesPoint {
  /// Time bucket
  bucket: string
  /// Timeout error count
  timeout_count: number
  /// Rate limit error count
  rate_limit_count: number
  /// Upstream 5xx error count
  upstream_5xx_count: number
  /// Circuit breaker trip count
  circuit_breaker_count: number
}

/// Error type breakdown for the Errors tab bar chart.
export interface ErrorTypeBreakdown {
  /// Error type
  error_type: string
  /// Count of errors
  count: number
  /// Percentage of total errors
  percentage: number
}

/// Error log row for the Errors tab table.
export interface ErrorLogRow {
  /// Request ID
  request_id: string
  /// Token ID
  token_id: string
  /// Token name
  token_name: string | null
  /// Model used
  model: string | null
  /// Error type
  error_type: string | null
  /// HTTP status from upstream
  upstream_status: number | null
  /// Response latency in ms
  response_latency_ms: number | null
  /// Deny reason
  deny_reason: string | null
  /// Timestamp
  created_at: string
}

File: gateway/src/models/analytics.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct VolumeStat {
    pub bucket: DateTime<Utc>,
    pub count: i64,
}

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct StatusStat {
    pub status_class: i32,
    pub count: i64,
}

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct LatencyStat {
    pub p50: f64,
    pub p90: f64,
    pub p99: f64,
    pub avg: f64,
}

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct TokenUsageBucket {
    pub bucket: DateTime<Utc>,
    pub count: i64,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TokenUsageStats {
    pub total_requests: i64,
    pub success_count: i64,
    pub error_count: i64,
    pub avg_latency_ms: f64,
    pub total_cost_usd: f64,
    pub hourly: Vec<TokenUsageBucket>,
}

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct AnalyticsSummary {
    pub total_requests: i64,
    pub success_count: i64,
    pub error_count: i64,
    pub avg_latency: f64,
    pub total_cost: f64,
    pub total_tokens: i64,
}

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct AnalyticsTimeseriesPoint {
    pub bucket: DateTime<Utc>,
    pub request_count: i64,
    pub error_count: i64,
    pub cost: f64,
    pub lat: f64,
}

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct ExperimentSummary {
    pub experiment_name: String,
    pub variant_name: String,
    pub total_requests: i64,
    pub avg_latency_ms: f64,
    pub total_cost_usd: f64,
    pub avg_tokens: f64,
    pub error_count: i64,
}

// ── Provider Analytics Types ──────────────────────────────────────

/// Model usage breakdown for analytics dashboard
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct ModelUsageStat {
    pub model: String,
    pub request_count: i64,
    pub cost_usd: f64,
}

/// Spend by provider for analytics dashboard
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct ProviderSpendStat {
    pub provider: String,
    pub spend_usd: f64,
    pub rate_per_1k: f64,
}

/// Latency by provider for analytics dashboard
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct ProviderLatencyStat {
    pub provider: String,
    pub latency_ms: f64,
}

// ── Traffic Analytics Types ──────────────────────────────────────

/// Traffic timeseries point with status breakdown for the Traffic tab.
/// Provides counts by policy_result category: passed, blocked, throttled, hitl-paused.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct TrafficTimeseriesPoint {
    pub bucket: DateTime<Utc>,
    pub total_count: i64,
    pub passed_count: i64,
    pub throttled_count: i64,
    pub blocked_count: i64,
    pub hitl_paused_count: i64,
}

/// Latency timeseries point with percentile breakdown for the Traffic tab.
/// Provides p50, p90, p99 latency values over time buckets.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct LatencyTimeseriesPoint {
    pub bucket: DateTime<Utc>,
    pub p50: f64,
    pub p90: f64,
    pub p99: f64,
}

// ── Cost Analytics Types ──────────────────────────────────────

/// Budget health status for the alert strip
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct BudgetHealthStatus {
    pub tokens_above_80_percent: i64,
    pub tokens_without_cap: i64,
    pub total_tokens: i64,
}

/// Spend timeseries grouped by a dimension (provider/model/token)
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct SpendTimeseriesPoint {
    pub bucket: DateTime<Utc>,
    pub dimension: String,
    pub spend_usd: f64,
    pub request_count: i64,
}

/// Cost efficiency by model over time
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct CostEfficiencyPoint {
    pub bucket: DateTime<Utc>,
    pub model: String,
    pub cost_per_1k_tokens: f64,
}

/// Budget burn rate calculation
#[derive(Debug, Serialize, Deserialize)]
pub struct BudgetBurnRate {
    pub days_elapsed: i32,
    pub days_remaining: i32,
    pub budget_usd: f64,
    pub spent_usd: f64,
    pub percent_used: f64,
    pub needed_per_day: f64,
    pub actual_per_day: f64,
    pub on_track: bool,
}

/// Enhanced spend breakdown with cap info
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct TokenSpendWithCap {
    pub token_id: String,
    pub token_name: String,
    pub provider: String,
    pub total_spend_usd: f64,
    pub spend_cap_usd: Option<f64>,
    pub percent_cap_used: Option<f64>,
    pub request_count: i64,
    pub cost_per_1k: f64,
}

// ── Cache Analytics Types ──────────────────────────────────────

/// Cache summary statistics for the Cache tab ribbon.
#[derive(Debug, Serialize, Deserialize)]
pub struct CacheSummaryStats {
    /// Overall cache hit rate as percentage (0-100)
    pub hit_rate: f64,
    /// Estimated cost avoided by cache hits (USD)
    pub cost_avoided_usd: f64,
    /// Total cache size in bytes (from Redis INFO)
    pub cache_size_bytes: i64,
    /// Most frequently cached model
    pub top_cached_model: Option<String>,
}

/// Cache hit rate timeseries point for the Cache tab chart.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct CacheHitRatePoint {
    pub bucket: DateTime<Utc>,
    pub hit_rate: f64,
    pub hit_count: i64,
    pub total_count: i64,
}

/// Top cached query entry for the Cache tab table.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct CachedQueryRow {
    /// Hash of the cache key (truncated for display)
    pub query_hash: String,
    /// Model associated with this cached query
    pub model: Option<String>,
    /// Number of cache hits for this query
    pub hits: i64,
    /// Time since last cache hit (human readable, calculated in frontend)
    pub last_hit_at: DateTime<Utc>,
    /// Age of the cache entry in seconds
    pub cache_age_seconds: i64,
    /// Seconds until cache expires (TTL remaining)
    pub expires_in_seconds: Option<i64>,
}

/// Model-level cache efficiency for the Cache tab.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct ModelCacheEfficiency {
    pub model: String,
    /// Cache hit rate for this model (0-100)
    pub hit_rate: f64,
    pub total_requests: i64,
    pub cache_hits: i64,
}

/// Cache latency comparison for the Cache tab.
#[derive(Debug, Serialize, Deserialize)]
pub struct CacheLatencyComparison {
    /// Average latency for cached responses (ms)
    pub cached_latency_ms: f64,
    /// Average latency for uncached responses (ms)
    pub uncached_latency_ms: f64,
    /// Speedup factor (uncached / cached)
    pub speedup_factor: f64,
    /// Sample size for cached responses
    pub cached_sample_count: i64,
    /// Sample size for uncached responses
    pub uncached_sample_count: i64,
}

// ── Model Analytics Types (Models Tab) ──────────────────────────────────────

/// Model usage timeseries point for the Models tab.
/// Used by the Model Usage Over Time chart.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct ModelUsageTimeseriesPoint {
    pub bucket: DateTime<Utc>,
    pub model: String,
    pub value: f64,
}

/// Error rate by model for the Models tab.
/// Used by the Error Rate by Model horizontal bar chart.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct ModelErrorRate {
    pub model: String,
    pub error_rate: f64,
    pub total_requests: i64,
    pub error_count: i64,
}

/// Latency by model for the Models tab.
/// Used by the Model Stats table and scatter chart.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct ModelLatencyStat {
    pub model: String,
    pub avg_latency_ms: f64,
    pub p50: f64,
    pub p90: f64,
    pub p99: f64,
    pub sample_count: i64,
}

/// Combined model stats for the Models tab table.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct ModelStatsRow {
    pub model: String,
    pub total_tokens: i64,
    pub total_cost_usd: f64,
    pub error_rate: f64,
    pub avg_latency_ms: f64,
    pub request_count: i64,
}

/// Cost vs latency scatter point for the Models tab.
/// Used by the Cost vs Latency bubble chart with proportional dot sizing.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct CostLatencyScatterPoint {
    pub model: String,
    pub avg_cost_per_request: f64,
    pub avg_latency_ms: f64,
    pub total_spend_usd: f64,
    pub total_requests: i64,
}

/// A/B test lift metrics for the Models tab.
/// Shows champion vs challenger comparison.
#[derive(Debug, Serialize, Deserialize)]
pub struct ABTestLiftResponse {
    pub experiment_id: String,
    pub experiment_name: String,
    pub champion: ABTestVariant,
    pub challenger: ABTestVariant,
    pub cost_delta_percent: f64,
    pub quality_metrics: Vec<QualityMetricDelta>,
}

/// A/B test variant info.
#[derive(Debug, Serialize, Deserialize)]
pub struct ABTestVariant {
    pub model: String,
    pub weight_percent: f64,
}

/// Quality metric delta for A/B test lift.
#[derive(Debug, Serialize, Deserialize)]
pub struct QualityMetricDelta {
    pub metric: String,
    pub delta: String,
}

// ── Security Analytics Types (Security Tab) ──────────────────────────────────────

/// Security KPI summary for the Security tab ribbon.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct SecuritySummaryStats {
    /// Total PII redactions (records with fields_redacted)
    pub pii_redactions: i64,
    /// Total guardrail blocks (Deny results)
    pub guardrail_blocks: i64,
    /// Total shadow mode violations
    pub shadow_violations: i64,
    /// External guardrail blocks (deny_reason contains 'external')
    pub external_blocks: i64,
}

/// Guardrail trigger count by category for the Security tab.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct GuardrailTriggerStat {
    /// Guardrail category (jailbreak, harmful, injection, profanity, bias, sensitive, competitor)
    pub category: String,
    /// Number of triggers for this category
    pub count: i64,
}

/// PII breakdown by pattern type for the Security tab.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct PiiBreakdownStat {
    /// PII pattern type (email, api_key, cc, ssn, phone, nlp)
    pub pattern: String,
    /// Number of redactions for this pattern
    pub count: i64,
}

/// Policy action count for the Security tab.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct PolicyActionStat {
    /// Policy action type (redact, content_filter, deny, rate_limit, require_approval, shadow)
    pub action: String,
    /// Number of times this action was applied
    pub count: i64,
}

/// Shadow mode policy with violation stats for the Security tab.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct ShadowPolicyStat {
    /// Policy name
    pub policy_name: String,
    /// Number of shadow violations
    pub violations: i64,
    /// Token with most violations
    pub top_token: String,
    /// Policy status (Monitoring/Would block)
    pub status: String,
}

/// Data residency stats for the Security tab.
#[derive(Debug, Serialize, Deserialize)]
pub struct DataResidencyStats {
    /// Percentage of EU-routed requests
    pub eu_percent: f64,
    /// Percentage of US-routed requests
    pub us_percent: f64,
}

// ── HITL Analytics Types (HITL Tab) ──────────────────────────────────────

/// HITL KPI summary for the HITL tab ribbon.
#[derive(Debug, Serialize, Deserialize)]
pub struct HitlSummaryStats {
    /// Number of pending approval requests
    pub pending_count: i64,
    /// Average wait time in seconds for completed approvals
    pub avg_wait_seconds: f64,
    /// Approval rate as percentage (0-100)
    pub approval_rate: f64,
}

/// HITL volume timeseries point for the HITL tab chart.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct HitlVolumePoint {
    /// Time bucket
    pub bucket: chrono::DateTime<chrono::Utc>,
    /// Count of approved requests
    pub approved_count: i64,
    /// Count of rejected requests
    pub rejected_count: i64,
    /// Count of expired requests (proxy for escalated)
    pub expired_count: i64,
    /// Count of pending requests
    pub pending_count: i64,
}

/// HITL latency stats for the SLA card.
#[derive(Debug, Serialize, Deserialize)]
pub struct HitlLatencyStats {
    /// P50 approval time in seconds
    pub p50_seconds: f64,
    /// P90 approval time in seconds
    pub p90_seconds: f64,
    /// P99 approval time in seconds
    pub p99_seconds: f64,
    /// Average approval time in seconds
    pub avg_seconds: f64,
}

/// Rejection reason breakdown for the HITL tab.
/// Note: rejection_reason column does not exist yet, returning mock data.
#[derive(Debug, Serialize, Deserialize)]
pub struct RejectionReason {
    /// Reason for rejection
    pub reason: String,
    /// Percentage of total rejections
    pub percentage: f64,
}

// ── Error Analytics Types (Errors Tab) ──────────────────────────────────────

/// Error KPI summary for the Errors tab ribbon.
#[derive(Debug, Serialize, Deserialize)]
pub struct ErrorSummaryStats {
    /// Total error count (upstream_status >= 400 OR upstream_status IS NULL)
    pub total_errors: i64,
    /// Error rate as percentage (0-100)
    pub error_rate: f64,
    /// Circuit breaker trip count (deny_reason = 'circuit_breaker_open')
    pub circuit_breaker_trips: i64,
    /// Rate limit hit count (error_type = 'rate_limit')
    pub rate_limit_hits: i64,
    /// Most common error type
    pub top_error_type: Option<String>,
    /// Prior period total errors for delta
    pub prior_total_errors: i64,
    /// Prior period total requests for delta
    pub prior_total_requests: i64,
}

/// Error timeseries point for the Errors tab chart.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct ErrorTimeseriesPoint {
    /// Time bucket
    pub bucket: chrono::DateTime<chrono::Utc>,
    /// Timeout error count
    pub timeout_count: i64,
    /// Rate limit error count
    pub rate_limit_count: i64,
    /// Upstream 5xx error count
    pub upstream_5xx_count: i64,
    /// Circuit breaker trip count
    pub circuit_breaker_count: i64,
}

/// Error type breakdown for the Errors tab bar chart.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct ErrorTypeBreakdown {
    /// Error type (rate_limit, timeout, etc.)
    pub error_type: String,
    /// Count of errors
    pub count: i64,
    /// Percentage of total errors
    pub percentage: f64,
}

/// Error log row for the Errors tab table.
#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct ErrorLogRow {
    /// Request ID
    pub request_id: String,
    /// Token ID
    pub token_id: String,
    /// Token name (from JOIN)
    pub token_name: Option<String>,
    /// Model used
    pub model: Option<String>,
    /// Error type
    pub error_type: Option<String>,
    /// HTTP status from upstream
    pub upstream_status: Option<i16>,
    /// Response latency in ms
    pub response_latency_ms: Option<i32>,
    /// Deny reason
    pub deny_reason: Option<String>,
    /// Timestamp
    pub created_at: chrono::DateTime<chrono::Utc>,
}
